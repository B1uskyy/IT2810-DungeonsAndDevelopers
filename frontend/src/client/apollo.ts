import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
