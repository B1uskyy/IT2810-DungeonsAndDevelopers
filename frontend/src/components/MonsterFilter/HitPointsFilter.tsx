import React, { useEffect, useState, useCallback } from 'react';
import { Slider, Box } from '@mui/material';

interface HitPointsFilterProps {
  minHp: number;
  maxHp: number;
  onHpRangeChange: (range: [number, number]) => void;
}

const HitPointsFilter: React.FC<HitPointsFilterProps> = ({ minHp, maxHp, onHpRangeChange }) => {
  const [hpRange, setHpRange] = useState<[number, number]>([minHp, maxHp]);

  useEffect(() => {
    setHpRange([minHp, maxHp]);
  }, [minHp, maxHp]);

  const handleHpChange = useCallback(
    (event: Event, newValue: number | number[]) => {
      if (Array.isArray(newValue)) {
        setHpRange(newValue as [number, number]);
        onHpRangeChange(newValue as [number, number]);
      }
    },
    [onHpRangeChange]
  );

  return (
    <Box>
      <Slider
        value={hpRange}
        onChange={handleHpChange}
        min={minHp}
        max={maxHp}
        step={1}
        valueLabelDisplay="auto"
        sx={{
          '& .MuiSlider-thumb': { color: '#DB3232', width: 24, height: 24 },
          '& .MuiSlider-track': { color: '#DB3232', height: 10 },
          '& .MuiSlider-rail': { color: '#DB3232', height: 10 },
        }}
      />
      <Box display="flex" width="300px" justifyContent="space-between" mt={1}>
        <span>{minHp}</span>
        <span>{maxHp}</span>
      </Box>
    </Box>
  );
};

export default HitPointsFilter;
