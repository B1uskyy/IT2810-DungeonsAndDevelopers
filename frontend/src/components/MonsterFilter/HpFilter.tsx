import { Box, Slider } from '@mui/material';
import { useState } from 'react';

interface HitPointsFilterProps {
  minHp: number;
  maxHp: number;
  onHpChange: (min: number, max: number) => void;
}

export default function HitPointsFilter({ minHp, maxHp, onHpChange }: HitPointsFilterProps) {
  const [hpRange, setHpRange] = useState<number[]>([minHp, maxHp]);

  const handleSliderChange = (_: Event, newValue: number | number[]) => {
    const [min, max] = newValue as number[];
    setHpRange([min, max]);
    onHpChange(min, max);
  };

  return (
    <Box>
      <Slider
        value={hpRange}
        onChange={handleSliderChange}
        valueLabelDisplay="auto"
        min={0}
        max={1000}
        sx={{
          '& .MuiSlider-thumb': { color: '#DB3232', width: 24, height: 24 },
          '& .MuiSlider-track': { color: '#DB3232', height: 10 },
          '& .MuiSlider-rail': { color: '#DB3232', height: 10 },
        }}
      />
      <Box display="flex" width="300px" justifyContent="space-between" mt={1}>
        <span>{hpRange[0]}</span>
        <span>{hpRange[1]}</span>
      </Box>
    </Box>
  );
}
