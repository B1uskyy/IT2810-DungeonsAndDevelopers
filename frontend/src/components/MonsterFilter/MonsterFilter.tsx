import { useState } from 'react';

const filterOptions = [
  'dragon',
  'aberration',
  'humanoid',
  'monstrosity',
  'beast',
  'plant',
  'construct',
  'fiend',
  'fey',
  'giant',
  'undead',
  'celestial',
  'elemental',
];

export default function MonsterFilter() {
  const [selectedFilters, setSelectedFilters] = useState<Set<string>>(new Set());
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);

  const handleCheckboxChange = (option: string) => {
    setSelectedFilters((prev) => {
      const newFilters = new Set(prev);
      if (newFilters.has(option)) {
        newFilters.delete(option);
      } else {
        newFilters.add(option);
      }
      return newFilters;
    });
  };

  const toggleDropdown = () => {
    setIsDropdownOpen((prev) => !prev);
  };

  return (
    <div className="relative text-white">
      <button onClick={toggleDropdown} className="bg-gray-800 p-2 rounded">
        Filter Monsters
      </button>
      {isDropdownOpen && (
        <div className="absolute bg-gray-700 p-4 rounded mt-2 w-48">
          {filterOptions.map((option) => (
            <label key={option} className="block">
              <input
                type="checkbox"
                checked={selectedFilters.has(option)}
                onChange={() => handleCheckboxChange(option)}
                className="mr-2"
              />
              {option}
            </label>
          ))}
        </div>
      )}
    </div>
  );
}
