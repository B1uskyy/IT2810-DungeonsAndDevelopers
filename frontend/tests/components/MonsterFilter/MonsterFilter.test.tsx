import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MonsterFilter from '../../../src/components/MonsterFilter/MonsterFilter.tsx';
import { describe, it, expect, vi, afterEach } from 'vitest';
import { MockedProvider } from '@apollo/client/testing';
import { AuthContext } from '../../../src/context/AuthContext.tsx';
import { GET_MONSTER_HP_RANGE } from '../../../src/graphql/getMonsterQuerie.ts';

// Mock props for the MonsterFilter component
const mockSetSelectedFilters = vi.fn();
let selectedFilters = new Set<string>();
const mockMonsterCounts = {
  dragon: 5,
  fiend: 0,
};
const mockOnClearFilters = () => {};
const mockOnHpChange = () => {};
const mockSetCurrentPage = () => {};

const mockUserId = 'user-321';
const mockToken = 'mock-token';
const mockUsername = 'Bob';
const mockLogin = vi.fn();
const mockLogout = vi.fn();

const mocks = [
  {
    request: {
      query: GET_MONSTER_HP_RANGE,
    },
    result: {
      data: {
        monsterHpRange: {
          minHp: 10,
          maxHp: 50,
          __typename: 'MonsterHpRange',
        },
      },
    },
  },
];
describe('MonsterFilter Component', () => {
  afterEach(() => {
    vi.restoreAllMocks();
    selectedFilters = new Set();
  });

  it('renders correctly and matches snapshot', () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider
          value={{ userId: mockUserId, userName: mockUsername, logout: mockLogout, login: mockLogin, token: mockToken }}
        >
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('opens and closes the dropdown menu', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider
          value={{ userId: mockUserId, userName: mockUsername, logout: mockLogout, login: mockLogin, token: mockToken }}
        >
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );
    const filterButton = screen.getByText('Filter Monsters');

    // Click to open dropdown
    await userEvent.click(filterButton);

    await waitFor(() => {
      expect(screen.getByText('Clear Filters')).toBeInTheDocument();
    });

    // Click on close icon to close dropdown
    const closeButton = screen.getByLabelText('Close');
    await userEvent.click(closeButton);
    await waitFor(() => {
      expect(screen.queryByText('Clear Filters')).not.toBeInTheDocument();
    });
  });

  it('toggles filters and updates selected filters correctly', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider
          value={{ userId: mockUserId, userName: mockUsername, logout: mockLogout, login: mockLogin, token: mockToken }}
        >
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );
    const filterButton = screen.getByText('Filter Monsters');

    // Open dropdown
    await userEvent.click(filterButton);
    await waitFor(() => {
      expect(screen.getByText('dragon (5)')).toBeInTheDocument();
    });

    // Select "dragon" filter
    const dragonLabel = screen.getByText('dragon (5)');
    await userEvent.click(dragonLabel);
    await waitFor(() => {
      expect(mockSetSelectedFilters).toHaveBeenCalledTimes(1);
      const updaterFunction = mockSetSelectedFilters.mock.calls[0][0];
      const updatedSet = updaterFunction(new Set());
      expect(updatedSet).toContain('dragon');
    });

    // Unselect "dragon" filter
    await userEvent.click(dragonLabel);
    await waitFor(() => {
      expect(mockSetSelectedFilters).toHaveBeenCalledTimes(2);
      const updaterFunction = mockSetSelectedFilters.mock.calls[1][0];
      const updatedSet = updaterFunction(new Set(['dragon']));
      expect(updatedSet).not.toContain('dragon');
    });
  });

  it('clears all selected filters', async () => {
    selectedFilters = new Set(['dragon', 'humanoid']);
    render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider
          value={{ userId: mockUserId, userName: mockUsername, logout: mockLogout, login: mockLogin, token: mockToken }}
        >
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );

    // Open dropdown
    const filterButton = screen.getByText('Filter Monsters');
    await userEvent.click(filterButton);

    // Click clear filters
    const clearButton = screen.getByText('Clear Filters');
    await userEvent.click(clearButton);

    expect(mockSetSelectedFilters).toHaveBeenCalledTimes(1);
    expect(mockSetSelectedFilters).toHaveBeenCalledWith(new Set());
  });
});
