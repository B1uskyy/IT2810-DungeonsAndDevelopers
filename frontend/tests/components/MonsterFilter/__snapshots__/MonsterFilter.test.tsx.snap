// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MonsterFilter Component > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="relative text-white"
  >
    <button
      class="text px-1 rounded-md bg-customRed hover:bg-transparent border-2 border-customRed hover:border-customRed hover:text-customRed transition-colors duration-200"
    >
      Filter Monsters
    </button>
  </div>
</div>
`;
